jax/experimental/array_api/_data_type_functions.py

```diff
@@ -15,120 +15,32 @@
 from jax._src import dtypes as _dtypes
-from jax.experimental.array_api._dtypes import (
-  bool, int8, int16, int32, int64, uint8, uint16, uint32, uint64,
-  float32, float64, complex64, complex128
-)
-
-_valid_dtypes = {
-    bool, int8, int16, int32, int64, uint8, uint16, uint32, uint64,
-    float32, float64, complex64, complex128
-}
-
-_promotion_table = {
-  (bool, bool): bool,
-  (int8, int8): int8,
-  (int8, int16): int16,
-  (int8, int32): int32,
-  (int8, int64): int64,
-  (int8, uint8): int16,
-  (int8, uint16): int32,
-  (int8, uint32): int64,
-  (int16, int8): int16,
-  (int16, int16): int16,
-  (int16, int32): int32,
-  (int16, int64): int64,
-  (int16, uint8): int16,
-  (int16, uint16): int32,
-  (int16, uint32): int64,
-  (int32, int8): int32,
-  (int32, int16): int32,
-  (int32, int32): int32,
-  (int32, int64): int64,
-  (int32, uint8): int32,
-  (int32, uint16): int32,
-  (int32, uint32): int64,
-  (int64, int8): int64,
-  (int64, int16): int64,
-  (int64, int32): int64,
-  (int64, int64): int64,
-  (int64, uint8): int64,
-  (int64, uint16): int64,
-  (int64, uint32): int64,
-  (uint8, int8): int16,
-  (uint8, int16): int16,
-  (uint8, int32): int32,
-  (uint8, int64): int64,
-  (uint8, uint8): uint8,
-  (uint8, uint16): uint16,
-  (uint8, uint32): uint32,
-  (uint8, uint64): uint64,
-  (uint16, int8): int32,
-  (uint16, int16): int32,
-  (uint16, int32): int32,
-  (uint16, int64): int64,
-  (uint16, uint8): uint16,
-  (uint16, uint16): uint16,
-  (uint16, uint32): uint32,
-  (uint16, uint64): uint64,
-  (uint32, int8): int64,
-  (uint32, int16): int64,
-  (uint32, int32): int64,
-  (uint32, int64): int64,
-  (uint32, uint8): uint32,
-  (uint32, uint16): uint32,
-  (uint32, uint32): uint32,
-  (uint32, uint64): uint64,
-  (uint64, uint8): uint64,
-  (uint64, uint16): uint64,
-  (uint64, uint32): uint64,
-  (uint64, uint64): uint64,
-  (float32, float32): float32,
-  (float32, float64): float64,
-  (float32, complex64): complex64,
-  (float32, complex128): complex128,
-  (float64, float32): float64,
-  (float64, float64): float64,
-  (float64, complex64): complex128,
-  (float64, complex128): complex128,
-  (complex64, float32): complex64,
-  (complex64, float64): complex128,
-  (complex64, complex64): complex64,
-  (complex64, complex128): complex128,
-  (complex128, float32): complex128,
-  (complex128, float64): complex128,
-  (complex128, complex64): complex128,
-  (complex128, complex128): complex128,
-}
-
-
-def _is_valid_dtype(t):
-  try:
-    return t in _valid_dtypes
-  except TypeError:
-    return False
-
-
-def _promote_types(t1, t2):
-  if not _is_valid_dtype(t1):
-    raise ValueError(f"{t1} is not a valid dtype")
-  if not _is_valid_dtype(t2):
-    raise ValueError(f"{t2} is not a valid dtype")
-  if result := _promotion_table.get((t1, t2), None):
-    return result
-  else:
-    raise ValueError("No promotion path for {t1} & {t2}")
-
 
+# TODO(micky774): Update jax.numpy dtypes to dtype *objects*
```

Comment:
What about by using a generic wrapper for the dtypes, e.g.
```python
from typing import TypeVar, Generic
import numpy as np

T = TypeVar("T", bound=np.dtype)
class JaxDType(Generic[T]):
    def __init__(self, dtype):
        self.dtype = dtype

    def __call__(self, *args, **kwargs):
        return self.dtype(*args, **kwargs)

    def __repr__(self):
        return f"JaxDType({self.dtype})"


float32 = JaxDType(np.float32)

print(float32(1.5)) # 1.5
print(float32(2.5).dtype) # dtype('float32')
print(repr(float32)) # JaxDType(<class 'numpy.float32'>)
```

Reply:
We already have that wrapper, it's called `_ScalarMeta`.

Reply:
Currently they're annotated as `Any`, and it's quite difficult to change that without large numbers of breakages because of how they're used by downstream libraries

Reply:
I think actually that should be fine for the array API -- I've updated the `isdtype` function to implicitly check for the `_ScalarMeta`-constructed types (didn't feel like figuring out that circ. dependency if it can be avoided). This enables the `jax.numpy` dtypes to satisfy the array API in how they can be used.